### PR TITLE
feat: add support for service account impersonation

### PR DESCRIPTION
**Pull request by enocom**
This is a port of https://github.com/GoogleCloudPlatform/cloud-sql-proxy/pull/1460.

### CODE REVIEW

**Review by kurtisvg — APPROVED** [review]

File: internal/proxy/proxy.go [hunk 1]
```diff
@@ -107,6 +110,83 @@ type Config struct {
...
+	ImpersonateDelegates []string
+}
+
+func (c *Config) credentialsOpt(l alloydb.Logger) (alloydbconn.Option, error) {
+	// If service account impersonation is configured, set up an impersonated
+	// credentials token source.
+	if c.ImpersonateTarget != "" {
+		var iopts []option.ClientOption
+		switch {
+		case c.Token != "":
```
Comment (hessjcg) [inline internal/proxy/proxy.go:130]: Seems like this switch statement from line 130-150 are much identical to lines 166-189. Could you somehow
combine those into a separate function?
```
iopts := c.withCredentialSource(l) // switch statement in here
if c.ImpersonateTarget != "" {
   ts, err := impersonate.CredentialsTokenSource( ... iopts)
   // ... error handling, etc
   return alloydbconn.WithTokenSource(ts)
}
return iopts
```

**Review by hessjcg — DISMISSED** [review]
See if you can get rid of the redundant switch statement.

File: internal/proxy/proxy.go (hunk 1 above)
Comment (kurtisvg) [reply internal/proxy/proxy.go:130]: If we are making significant changes to this PR we should make sure to update the Cloud SQL go connector as well

Comment (enocom) [reply internal/proxy/proxy.go:130]: Let's fix this in the v2 proxy and port the change here.

**Pull request merged**